source/vsignal/signal.d

```diff
@@ -0,0 +1,112 @@
+module vsignal.signal;
+
+import vsignal.sink;
+import vsignal.slot;
+
+struct Signal(F)
+	if (is(F : RT delegate(Args), RT, Args...))
+{
+	/**
+	Publish a signal.
+
+	Params:
+		args = the Signal's function parameters to call each listener with.
+	*/
+	void emit(Slot!F.Parameters args)
+	{
+		import core.lifetime : forward;
+
+		foreach (call; calls)
+			cast(void) call(forward!args);
+	}
+
+	bool empty() const @property
+	{
+		return !calls.length;
+	}
+
+	size_t length() const @property
+	{
+		return calls.length;
+	}
+
+	Sink!F sink() @property
+	{
+		return Sink!F(&this);
+	}
+
+package:
+	Slot!F[] calls;
+}
+
+///
+@safe pure nothrow unittest
+{
+	Signal!(void delegate(int) @safe pure nothrow) signal;
+	Signal!(void delegate(ref int) @safe pure nothrow) signalRef;
+
+	static struct Listener
+	{
+		int i;
+		void opCall(const ref int)
+		{
+			i++;
+		}
+	}
+
+	Listener listener;
+
+	signal.sink.connect!(Listener.opCall)(listener);
+	signalRef.sink.connect!(Listener.opCall)(listener);
+
+	int var;
+
+	signal.emit(var);
+	signalRef.emit(var);
+
+	assert(listener.i == 2);
+}
+
+///
+@safe nothrow unittest
+{
+	Signal!(void delegate(int) @safe nothrow @nogc) signal;
+	int var = 35;
+
+	signal.sink.connect!((ref i, const int x) @trusted {
```

Comment:
Considering that this is an example `unittest`, I wouldn't show usage of @trusted code to the end user.

Reply:
It's @trusted because you cannot access the address of a `ref` in @safe core. I can move the @trusted to the assert call.

Reply:
I would make this a non-example `unittest` and write one without that check for documentation purposes.